internal/cluster/transition.go

```diff
@@ -153,3 +161,124 @@ func (th *transitionHandler) updateTransitionCluster(log logrus.FieldLogger, db
+	metricApi         metrics.API
+	hostApi           host.API
+	conditions        map[validationID]bool
+	validationResults map[string][]validationResult
+	db                *gorm.DB
+}
+
+func If(id validationID) stateswitch.Condition {
+	ret := func(sw stateswitch.StateSwitch, args stateswitch.TransitionArgs) (bool, error) {
+		params, ok := args.(*TransitionArgsRefreshCluster)
+		if !ok {
+			return false, errors.Errorf("If(%s) invalid argument", id.String())
+		}
+		b, ok := params.conditions[id]
+		if !ok {
+			return false, errors.Errorf("If(%s) no such condition", id.String())
+		}
+		return b, nil
+	}
+	return ret
+}
+
+// Return a post transition function with a constant reason
+func (th *transitionHandler) PostRefreshCluster(reason string) stateswitch.PostTransition {
+	ret := func(sw stateswitch.StateSwitch, args stateswitch.TransitionArgs) error {
+		sCluster, ok := sw.(*stateCluster)
+		if !ok {
+			return errors.New("PostRefreshCluster incompatible type of StateSwitch")
+		}
+		params, ok := args.(*TransitionArgsRefreshCluster)
+		if !ok {
+			return errors.New("PostRefreshCluster invalid argument")
+		}
+		var (
+			b              []byte
+			err            error
+			updatedCluster *common.Cluster
+		)
+		b, err = json.Marshal(&params.validationResults)
+		if err != nil {
+			return err
+		}
+		updatedCluster, err = updateClusterStatus(logutil.FromContext(params.ctx, th.log), params.db, *sCluster.cluster.ID, sCluster.srcState, *sCluster.cluster.Status,
+			reason, "validations_info", string(b))
+		//update hosts status to models.HostStatusResettingPendingUserAction if needed
+		cluster := sCluster.cluster
+		if updatedCluster != nil {
+			cluster = updatedCluster
+		}
+		setPendingUserResetIfNeeded(params.ctx, logutil.FromContext(params.ctx, th.log), params.db, params.hostApi, cluster)
+		//if status was changed - we need to send event and metrics
+		if err == nil && updatedCluster != nil && sCluster.srcState != swag.StringValue(updatedCluster.Status) {
+			msg := fmt.Sprintf("Updated status of cluster %s to %s", updatedCluster.Name, *updatedCluster.Status)
+			params.eventHandler.AddEvent(params.ctx, updatedCluster.ID.String(), models.EventSeverityInfo, msg, time.Now(), updatedCluster.ID.String())
+			//report installation finished metric if needed
+			reportInstallationCompleteStatuses := []string{models.ClusterStatusInstalled, models.ClusterStatusError}
+			if sCluster.srcState == models.ClusterStatusInstalling &&
+				funk.ContainsString(reportInstallationCompleteStatuses, swag.StringValue(updatedCluster.Status)) {
+				params.metricApi.ClusterInstallationFinished(logutil.FromContext(params.ctx, th.log), swag.StringValue(updatedCluster.Status),
+					updatedCluster.OpenshiftVersion, updatedCluster.InstallStartedAt)
+			}
+			return nil
+		}
+		return err
+	}
+	return ret
+}
+
+func setPendingUserResetIfNeeded(ctx context.Context, log logrus.FieldLogger, db *gorm.DB, hostApi host.API, c *common.Cluster) {
```

Comment:
Seems that this whole section regarding the "pending-user-reset" should be part of host refresh status, but not as host validations. This operation accesses hosts, but does not rely on any information from the cluster.

Reply:
I agree but there is a problem to do it - when the host gets into pending user reset - all hosts should be in the same state. since this status is not reflected in the cluster - this is the way to update all hosts - @RazRegev can explain better.
@RazRegev maybe we should open a task to fix this similar to what we did when the host gets into error state.

Reply:
We don't check here that all hosts are in the same state. We need just one. In addition, in the host state machine we just check that the host is in the state HostStatusResetting which is what is done here.  It is redundant.
Instead, it can be done as a single transition in the host state machine.